astropy/table/operations.py

```diff
@@ -697,6 +697,69 @@ def common_dtype(cols):
         raise tme
 
 
+def _get_join_sort_idxs(out_descrs, keys, len_left, len_right, left, right):
+
+    # Remake out_descrs as a dict keyed by the name (descr[0])
+    descrs = {descr[0]: descr for descr in out_descrs}
+
+    # Go through each of the key columns in order and make columns for
+    # a new structured array that represents the lexical ordering of those
+    # key columns. This structured array is then argsort'ed. The trick here
+    # is that some columns (e.g. Time) may need to be expanded into multiple
+    # columns for ordering here.
+
+    ii = 0  # Index for uniquely naming the sort columns
+    sort_keys_dtypes = []  # sortable_table dtypes as list of (name, dtype_str, shape) tuples
+    sort_keys = []  # sortable_table (structured ndarray) column names
+    sort_left = {}  # sortable ndarrays from left table
+    sort_right = {}  # sortable ndarray from right table
+
+    for key in keys:
+        # TODO: shape here seems out of place. Can a key column have a non-trivial shape?
+        name, _, shape = descrs[key]
+
+        # get_sortable_arrays() returns a list of ndarrays that can be lexically
+        # sorted to represent the order of the column. In most cases this is just
+        # a single element of the column itself.
+        left_sort_cols = left[name].info.get_sortable_arrays()
+        right_sort_cols = right[name].info.get_sortable_arrays()
+
+        if len(left_sort_cols) != len(right_sort_cols):
+            # Should never happen because cols are screened beforehand for compatibility
+            raise ValueError('unexpected mismatch in sort cols lengths')
```

Comment:
Maybe use `RuntimeError` for things that should never happen (still good to check).

Alternatively, I think it makes sense to use `assert ..., "<reason>"` here, so that it does not get executed with `python -O`.

Reply:
I went with `RuntimeError`.  I don't actually understand why you would want to not execute this test via `python -O`.  I must just be missing something.

Reply:
I don't think it matter much, since the test is far cheaper than the rest, but I think `assert` is reasonable since you write right about that this "Should never happen", i.e., you have already tested in earlier code and thus here you simply 'assert' that the earlier checks were done. (I like this use of assert - it asserts assumptions made in the code that should hold, but that, given that bugs do happen, may be wrong any way - I think when you optimize for speed, it is fine to go with the assumptions...)